### PR TITLE
Project canopy

### PR DESCRIPTION
**Pull request by jarry-xiao**
Project canopy increases the potential Merkle tree size by an order of magnitude by allowing Gummyroll to store more hashes on chain. The main problem that this solves is that it enables users to supply proofs while passing in fewer hashes into the transaction (which is limited to 1232 bytes).

This change can support the ability to index and operate on trees with significantly more leaves than was previously possible with the added benefit of not modifying any existing APIs

### CODE REVIEW

File: contracts/programs/bubblegum/src/lib.rs [hunk 1]
```diff
@@ -567,14 +566,15 @@ pub mod bubblegum {
         }?;
         let merkle_slab = ctx.accounts.merkle_slab.to_account_info();
         emit!(voucher.leaf_schema.to_event());
-        insert_or_append_leaf(
+        replace_leaf(
```
Comment (jarry-xiao) [inline contracts/programs/bubblegum/src/lib.rs:569]: I think it makes sense for `cancel_redeem` to only attempt to replace given that the nonce (index) is known

File: lib/concurrent-merkle-tree/src/merkle_roll.rs [hunk 2]
```diff
@@ -358,33 +322,30 @@ impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> MerkleRoll<MAX_DEPTH,
...
+    ) -> Result<bool, CMTError> {
         let mask: usize = MAX_BUFFER_SIZE - 1;
-        let changelog_buffer_index: u64;
-        if use_full_buffer {
-            changelog_buffer_index = self.active_index.wrapping_sub(self.buffer_size) & mask as u64
-        } else {
-            match self.find_root_in_changelog(current_root) {
-                Some(matching_changelog_index) => {
-                    changelog_buffer_index = matching_changelog_index;
+        let (changelog_buffer_index, use_full_buffer) = match self
```
Comment (jarry-xiao) [inline lib/concurrent-merkle-tree/src/merkle_roll.rs:335]: `use_full_buffer` should be determined, not specified

File: contracts/programs/gummyroll/src/lib.rs [hunk 3]
```diff
@@ -58,70 +60,196 @@ pub struct TransferAuthority<'info> {
...
+        // node_idx - 2 maps to the canopy index
+        let shifted_index = node_idx as usize - 2;
+        if shifted_index % 2 == 0 {
+            inferred_nodes.push(canopy[shifted_index + 1])
+        } else {
+            inferred_nodes.push(canopy[shifted_index - 1])
+        }
+        node_idx >>= 1;
+    }
+    proof.extend(inferred_nodes.iter());
```
Comment (jarry-xiao) [inline contracts/programs/gummyroll/src/lib.rs:160]: The correct behavior here is probably to only extend if the proof node is not already provided. Will add the change in soon

**Review by samwise2 — APPROVED** [review]
lgtm! super helpful for marketplace 🎉

**Pull request merged**